### PR TITLE
Autoresponses

### PR DESCRIPTION
**Pull request by JulianVennen**

### CODE REVIEW

**Review by KurtThiemann — APPROVED** [review]

File: commands/autoresponse.js [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+            }
+            id = parseInt(args.shift());
+            if (!responses[id]) {
+                await message.channel.send("Invalid id!");
+                return;
+            }
+            response = responses[id];
+
+            let confirmation = await message.channel.send("Do you really want to delete this response?",new Discord.MessageEmbed()
+                .setTitle("Added new auto-response")
```
Comment (matthi4s) [inline commands/autoresponse.js:138]: ?

File: commands/autoresponse.js [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+                return;
+            }
+            id = parseInt(args.shift());
+            if (!responses[id]) {
+                await message.channel.send("Invalid id!");
+                return;
+            }
+            response = responses[id];
+
+            await message.channel.send("Do you really want to delete this response?",new Discord.MessageEmbed()
```
Comment (matthi4s) [inline commands/autoresponse.js:180]: ?

File: commands/autoresponse.js [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+                .setTitle("Added new auto-response")
```
Comment (matthi4s) [inline commands/autoresponse.js:181]: ?

**Review by matthi4s — CHANGES_REQUESTED** [review]
In general, I'm not really happy with this long spaghetti code in the command and storing all auto-responses in the guild config is not really a good solution.

**Review by matthi4s — CHANGES_REQUESTED** [review]
I think at least the channel/global option should be a dedicated column. You can also only load the responses for the current channel then.

**Comment by JulianVennen** [comment]
Now the only thing that is stored as JSON is the trigger

**Comment by JulianVennen** [comment]
removing a response still refreshes the cache. Should I change that?

**Comment by matthi4s** [comment]
Just remove the response from cache?

**Comment by JulianVennen** [comment]
alright.
Anything else?

**Review by KurtThiemann — APPROVED** [review]

**Review by KurtThiemann — APPROVED** [review]

File: commands/autoresponse.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    //Permission check
+    if (!message.member.hasPermission('MANAGE_GUILD')) {
+        await message.channel.send('You need the "Manage Server" permission to use this command.');
+        return;
+    }
+    if (!args.length) {
+        return await message.channel.send(await util.usage(message,command.names[0]));
+    }
+
+    let responses = await util.getAutoResponses(message.channel.id, message.guild.id);
```
Comment (matthi4s) [inline commands/autoresponse.js:26]: You are loading channel-specific responses here, even though you should load all for the guild.

**Review by matthi4s — CHANGES_REQUESTED** [review]

File: commands/autoresponse.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        case 'add':
+            await add(responses, message);
+            break;
+
+        case 'delete':
+        case 'remove':
+            await remove(responses, message, args, database);
+            break;
+
+        case "info":
```
Comment (matthi4s) [inline commands/autoresponse.js:42]:
```suggestion
        case 'info':
```

File: commands/autoresponse/add.js [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    }
+    else {
+        options.global = false;
+        options.channels = await util.channelMentions(message.guild,channels.split(" "));
+    }
+
+    const response = new AutoResponse(message.guild.id, options);
+    response.id = await util.saveResponse(response);
+
+    await message.channel.send(util.responseEmbed(response,"Added new auto-response",util.color.green));
```
Comment (matthi4s) [inline commands/autoresponse/add.js:82]:
```suggestion
    await message.channel.send(util.responseEmbed(response,"Added new autoresponse",util.color.green));
```

File: commands/autoresponse/info.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+/**
+ * get info about an autoresponse
+ * @param {Object} responses
+ * @param {Discord.Message} message
+ * @param {String[]} args
+ * @returns {Promise<void>}
+ */
+module.exports = async (responses, message, args) => {
+    if (!args.length) {
+        await message.channel.send("Provide the id of the response you want to view");
```
Comment (matthi4s) [inline commands/autoresponse/info.js:13]:
```suggestion
        await message.channel.send("Provide the id of the autoresponse you want to view");
```

File: commands/autoresponse/info.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+        return;
+    }
+    let response = responses.get(parseInt(args.shift()));
+    if (!response) {
+        await message.channel.send("Invalid id!");
+        return;
+    }
+
+    await message.channel.send(util.responseEmbed(response,"Auto-response",util.color.green));
```
Comment (matthi4s) [inline commands/autoresponse/info.js:22]:
```suggestion
    await message.channel.send(util.responseEmbed(response,"Autoresponse",util.color.green));
```

File: commands/autoresponse/list.js [hunk 9]
```diff
@@ -0,0 +1,21 @@
+/**
+ * list auto responses
+ * @param {Object} responses
+ * @param {Discord.Message} message
+ * @returns {Promise<void>}
+ */
+module.exports = async (responses, message) => {
+    if (!responses.size) {
+        return await message.channel.send("No auto-responses!");
```
Comment (matthi4s) [inline commands/autoresponse/list.js:9]:
```suggestion
        return await message.channel.send("No autoresponses!");
```

File: commands/autoresponse/remove.js [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+ * remove an auto-response
+ * @param {Object} responses
+ * @param {Discord.Message} message
+ * @param {String[]} args
+ * @param {Database} database
+ * @returns {Promise<void>}
+ */
+module.exports = async (responses, message, args, database) => {
+    if (!args.length) {
+        await message.channel.send("Provide the id of the response you want to remove");
```
Comment (matthi4s) [inline commands/autoresponse/remove.js:14]:
```suggestion
        await message.channel.send("Provide the id of the autoresponse you want to remove");
```

File: commands/autoresponse/remove.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+        return;
+    }
+    let response = responses.get(parseInt(args.shift()));
+    if (!response) {
+        await message.channel.send("Invalid id!");
+        return;
+    }
+
+    let confirmation = await message.channel.send("Do you really want to delete this response?", util.responseEmbed(response,"Remove auto-response", util.color.red));
```
Comment (matthi4s) [inline commands/autoresponse/remove.js:23]:
```suggestion
    let confirmation = await message.channel.send("Do you really want to delete this autoresponse?", util.responseEmbed(response,"Remove autoresponse", util.color.red));
```

File: commands/autoresponse/remove.js [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    catch {
+        return await message.channel.send("You took to long to react!");
+    }
+    if (!confirmed) {
+        return await message.channel.send("Canceled!");
+    }
+
+    await util.removeResponse(response);
+
+    await message.channel.send(`Removed the response with the id ${response.id}!`);
```
Comment (matthi4s) [inline commands/autoresponse/remove.js:45]:
```suggestion
    await message.channel.send(`Removed the autoresponse with the id ${response.id}!`);
```

**Review by matthi4s — CHANGES_REQUESTED** [review]

**Review by matthi4s — APPROVED** [review]

**Review by KurtThiemann — APPROVED** [review]

**Pull request merged**